### PR TITLE
implement basic authorization module

### PR DESCRIPTION
**Pull request by kreuzhofer**
@tomkuijsten I have implemented the basic authorization. To activate it, you add an IAuthorizationProvider to the constructor of RestRouteHandler and StaticFileRouteHandler. So you can define which paths have authorization and which not. I would also like to add this to the wiki if you would make me a contributor of your project.
You have to pass an ICredentialValidator to the constructor of the BasicAuthorizationProvider. There you can implement username and password check.
Next I would implement another ICredentialValidator with pass through authentication that checks the username/password against the local administrator account, which is also used to log in to the admin interface of the pi.
Btw I did not yet write any unit tests but I tested locally and on the raspberry and it works fine.

### CODE REVIEW

**Comment by kreuzhofer** [comment]
Ok, this time it get errors. You will need to check the conflicts

**Comment by kreuzhofer** [comment]
@tomkuijsten I tried to resolve conflicts in github but every time I klick on the Button above I just get the spinning wheel an nothing happens. Strange.

**Comment by tomkuijsten** [comment]
Ill sort this out after the holidays, thanks for the PR!

File: scripts/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+        {
+            "type": "PowerShell",
+            "request": "launch",
+            "name": "PowerShell Launch (current file)",
+            "script": "${file}",
+            "args": [],
+            "cwd": "${file}"
+        }
+    ]
+}
```
Comment (tomkuijsten) [inline scripts/.vscode/launch.json:13]: This file should be removed right?

**Review by tomkuijsten — CHANGES_REQUESTED** [review]
Looks pretty solid, might refactor some stuff after the commit, but that will be more style preference then functional changes.

**Comment by Jark** [comment]
I think we should have unit tests before we merge it back though. 

We shouldn't add any features that are not at least tested in a basic manner, they become a pain to support in future when we want to make changes in the structure.

**Comment by kreuzhofer** [comment]
@tomkuijsten  Yes, Launch.json is not needed.
@Jark I agree, I will add some Basic unit Tests for my Code to help out

**Comment by kreuzhofer** [comment]
@Jark Could you please first do the merge so I can Switch over to the development branch for further changes. This will help to reduce the merge conflicts.

**Comment by Jark** [comment]
Hi @kreuzhofer,

I've created a new branch `feature-implement-basic-authorization-module` based off development and cherry-picked your change over, it's here -> https://github.com/tomkuijsten/restup/tree/feature-implement-basic-authorization-module

If you update your fork from the tomkuijsten repo and checkout that branch then you can add your unit tests and create a new pull request.

Alternatively you can try and rebase your change on top of development instead of master and the pull request should update as well.

Hope this helps.

Cheers,

Jark

**Comment by tomkuijsten** [comment]
I agree on the unit-tests, but I don't mind merging first. It's the development branch, so we won't release it to the public and still be able to remove it.

**Comment by kreuzhofer** [comment]
I have created a new pull request with both the changes and new unit tests

**Pull request closed without merging**